p4-16/spec/P4-16-spec.mdk

```diff
@@ -4444,6 +4469,62 @@ control c(...) {
 }
 ~ End P4Example
 
+### Instantiating objects with abstract methods {#sec-instantiating-abstract-methods}
+
+When instantiating an extern type that has `abstract` methods users
+have to supply implementations for all such methods.  This is done using object initializers:
+
+~ Begin P4Grammar
+lvalue:
+    ...
+    | THIS
+
+expression:
+    ...
+    | THIS
+
+instantiation:
+      ...
+      | annotations typeRef "(" argumentList ")" name "=" objInitializer ";"
+      | typeRef "(" argumentList ")" name "=" objInitializer ";"
+
+objInitializer
+    : "{" objDeclarations "}"
+    ;
+
+objDeclarations
+    : /* empty */
+    | objDeclarations objDeclaration
+    ;
+
+objDeclaration
+    : functionDeclaration
+    | instantiation
+    ;
+~ End P4Grammar
+
+The abstract methods can only use the supplied arguments or refer to
+values that are in the top-level scope.  When calling another method
```

Comment:
@cc10512 @ChrisDodd I have a half-memory that the primary interest in abstract functions was for ones that could read any variable/packet-header-field in its lexical scope, not only parameters and the top-level scope?

Perhaps I am remembering that about virtual functions, and I am not really clear right now on whether those are the same as, or different than, this definition of abstract functions.

Reply:
We discussed that we will introduce these features one by one.

Reply:
Yes, all the use cases we have for them involve being able to access any variable in scope at the point of the definition of the implementation method.